pxd.c

```diff
@@ -610,15 +623,15 @@ static int pxd_init_disk(struct pxd_device *pxd_dev, struct pxd_add_out *add)
 	disk->fops = &pxd_bd_ops;
 	disk->private_data = pxd_dev;
 
-	/* Bypass queue if queue_depth is zero. */
-	if (add->queue_depth == 0) {
-		q = blk_alloc_queue(GFP_KERNEL);
-		if (!q) {
-			err = -ENOMEM;
-			goto out_disk;
-		}
-		blk_queue_make_request(q, pxd_make_request);
```

Comment:
From what I can recollect, the reason we move away from bio based request processing is because we were running into deadlocks.

pxd_make_request gets called directly from generic_make_request from bdi_writeback threads. Whereas pxd_rq_fn is called when a particular pxd device's request queue is run (using a kthread). 
In the first case using bio submission:
- It doesn't honor the nr_request set on the queue. So bio get submitted unthrottled.
- And pxd_make_request can block, so the bdi thread would block. If u have num_device==bdi _threads potentially all can get locked.

Both of the above are addressed when using pxd_rq_fn. All this is wrt to 3.10 kernel, so unless something changed in the blocked device layer, we would again run into the old deadlock issue.

Reply:
The bdi writeback thread does not submit without checking for congestion, it uses the bdi->congestion_fn to query the backing device for accepting newer requests per inode over that backing device. This is being used by md/brbd etc to control congestion. I will extend the driver to setup this hook function as well.
 With the current congestion design, it blocks over a hard limit that would be useful for raw device access, but over filesystems, it shall control using the congestion fn, that will kick in earlier than the hard limit.

Reply:
Ok, so probably that takes care of issue#1.  Can you verify that this works both on kernel 3.10 and 4?
Issue#2 is still possible, if all bdi threads are processing pxd devices and they block on fuse mem allocation and most mem is held up in vfs for px devices.

Reply:
I think we had good conversion around this during our last call. bottom line needs more testing, and not to affect current io path. have addressed this under PR9.